package com.anthropic.models.beta.models

import com.anthropic.core.jsonMapper
import com.fasterxml.jackson.module.kotlin.jacksonTypeRef
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class BetaThinkingCapabilityTest {

    @Test
    fun create() {
        val betaThinkingCapability =
            BetaThinkingCapability.builder()
                .supported(true)
                .types(
                    BetaThinkingTypes.builder()
                        .adaptive(BetaCapabilitySupport.of(true))
                        .enabled(BetaCapabilitySupport.of(true))
                        .build()
                )
                .build()

        assertThat(betaThinkingCapability.supported()).isEqualTo(true)
        assertThat(betaThinkingCapability.types())
            .isEqualTo(
                BetaThinkingTypes.builder()
                    .adaptive(BetaCapabilitySupport.of(true))
                    .enabled(BetaCapabilitySupport.of(true))
                    .build()
            )
    }

    @Test
    fun roundtrip() {
        val jsonMapper = jsonMapper()
        val betaThinkingCapability =
            BetaThinkingCapability.builder()
                .supported(true)
                .types(
                    BetaThinkingTypes.builder()
                        .adaptive(BetaCapabilitySupport.of(true))
                        .enabled(BetaCapabilitySupport.of(true))
                        .build()
                )
                .build()

        val roundtrippedBetaThinkingCapability =
            jsonMapper.readValue(
                jsonMapper.writeValueAsString(betaThinkingCapability),
                jacksonTypeRef<BetaThinkingCapability>(),
            )

        assertThat(roundtrippedBetaThinkingCapability).isEqualTo(betaThinkingCapability)
    }
}
